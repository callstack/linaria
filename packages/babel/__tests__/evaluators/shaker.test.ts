import path from 'path';
import dedent from 'dedent';
import type { TransformOptions } from '@babel/core';
import shake from '../../../shaker/src';

function getFileName() {
  return path.resolve(__dirname, `../__fixtures__/test.js`);
}

function _shake(opts?: TransformOptions, only: string[] = ['__linariaPreval']) {
  return (
    literal: TemplateStringsArray,
    ...placeholders: string[]
  ): [string, Map<string, string[]>] => {
    const code = dedent(literal, ...placeholders);
    const [shaken, deps] = shake(
      getFileName(),
      {
        babelOptions: opts || {},
        displayName: true,
        evaluate: true,
        rules: [
          {
            action: require('../../../extractor/src').default,
          },
          {
            test: /\/node_modules\//,
            action: 'ignore',
          },
        ],
      },
      code,
      only
    );

    return [shaken, deps!];
  };
}

it('removes all', () => {
  const [shaken] = _shake()`
    const { whiteColor: color, anotherColor } = require('…');
    const a = color || anotherColor;
    color.green = '#0f0';

    exports.__linariaPreval = [];
  `;

  expect(shaken).toMatchSnapshot();
});

it('keeps only code which is related to `color`', () => {
  const [shaken] = _shake()`
    const { whiteColor: color, anotherColor } = require('…');
    const wrap = '';
    const a = color || anotherColor;
    color.green = '#0f0';
    module.exports = { color, anotherColor };
    exports.__linariaPreval = [color];
  `;

  expect(shaken).toMatchSnapshot();
});

it('keeps only code which is related to `anotherColor`', () => {
  const [shaken] = _shake()`
    const { whiteColor: color, anotherColor } = require('…');
    const a = color || anotherColor;
    color.green = '#0f0';
    exports.__linariaPreval = [anotherColor];
  `;

  expect(shaken).toMatchSnapshot();
});

it('keeps only code which is related to `a`', () => {
  const [shaken] = _shake()`
    const { whiteColor: color, anotherColor } = require('…');
    const a = color || anotherColor;
    color.green = '#0f0';
    exports.__linariaPreval = [a];
  `;

  expect(shaken).toMatchSnapshot();
});

it('shakes imports', () => {
  const [shaken] = _shake()`
    import { unrelatedImport } from '…';
    import { whiteColor as color, anotherColor } from '…';
    import defaultColor from '…';
    import anotherDefaultColor from '…';
    import '…';
    require('…');
    export default color;
    exports.__linariaPreval = [color, defaultColor];
  `;

  expect(shaken).toMatchSnapshot();
});

it('should keep member expression key', () => {
  const [shaken] = _shake()`
    const key = 'blue';
    const obj = { blue: '#00F' };
    const blue = obj[key];
    exports.__linariaPreval = [blue];
  `;

  expect(shaken).toMatchSnapshot();
});

it('shakes exports', () => {
  const [shaken] = _shake()`
    import { whiteColor as color, anotherColor } from '…';
    export const a = color;
    export { redColor } from "…";
    export { anotherColor };
    exports.__linariaPreval = [a];
  `;

  expect(shaken).toMatchSnapshot();
});

it('shakes es5 exports', () => {
  const [shaken] = _shake(undefined, ['redColor', 'greenColor', 'yellowColor'])`
    "use strict";
    Object.defineProperty(exports, "__esModule", {
      value: true
    });
    exports.redColor = 'red';
    exports['yellowColor'] = 'yellow';
    exports['pinkColor'] = 'pink';
    Object.defineProperty(exports, "blueColor", {
      enumerable: true,
      get: function get() {
        return 'blue';
      }
    });
    Object.defineProperty(exports, "greenColor", {
      enumerable: true,
      get: function get() {
        return 'green';
      }
    });
  `;

  expect(shaken).toMatchSnapshot();
});

// TODO: this test will be disabled until the shaker is fully implemented
// eslint-disable-next-line jest/no-disabled-tests
it.skip('should throw away any side effects', () => {
  const [shaken] = _shake()`
    const objects = { key: { fontSize: 12 } };
    const foo = (k) => {
      const obj = objects[k];
      console.log('side effect');
      return obj;
    };
    exports.__linariaPreval = [foo];
  `;

  expect(shaken).toMatchSnapshot();
});

it('keeps objects as is', () => {
  const [shaken] = _shake()`
    const fill1 = (top = 0, left = 0, right = 0, bottom = 0) => ({
      position: 'absolute',
      top,
      right,
      bottom,
      left,
    });

    const fill2 = (top = 0, left = 0, right = 0, bottom = 0) => {
      return {
        position: 'absolute',
        top,
        right,
        bottom,
        left,
      };
    };

    exports.__linariaPreval = [fill1, fill2];
  `;

  expect(shaken).toMatchSnapshot();
});

it('shakes sequence expression', () => {
  const [shaken] = _shake()`
    import { external } from '…';
    const color1 = (external, () => 'blue');
    let local = '';
    const color2 = (local = color1(), () => local);
    exports.__linariaPreval = [color2];
  `;

  expect(shaken).toMatchSnapshot();
});

it('shakes assignment patterns', () => {
  const [shaken] = _shake()`
    const [identifier = 1] = [2];
    const [{...object} = {}] = [{ a: 1, b: 2 }];
    const [[...array] = []] = [[1,2,3,4]];
    const obj = { member: null };
    ([obj.member = 42] = [1]);
    exports.__linariaPreval = [identifier, object, array, obj];
  `;

  expect(shaken).toMatchSnapshot();
});

it('shakes for-in statements', () => {
  const [shaken] = _shake()`
    const obj1 = { a: 1, b: 2 };
    const obj2 = {};
    for (const key in obj1) {
      obj2[key] = obj1[key];
    }
    exports.__linariaPreval = [obj2];
  `;

  expect(shaken).toMatchSnapshot();
});

it('keeps reused exports', () => {
  const [shaken] = _shake()`
    const bar = function() {
      return 'hello world';
    };
    exports.bar = bar;

    const foo = exports.bar();
    exports.__linariaPreval = [foo];
  `;

  expect(shaken).toMatchSnapshot();
});
