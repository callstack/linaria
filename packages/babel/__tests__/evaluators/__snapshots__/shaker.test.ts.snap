// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`keeps objects as is 1`] = `
"\\"use strict\\";

var fill1 = function fill1() {
  var top = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 0;
  var left = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 0;
  var right = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : 0;
  var bottom = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : 0;
  return {
    position: 'absolute',
    top: top,
    right: right,
    bottom: bottom,
    left: left
  };
};

var fill2 = function fill2() {
  var top = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 0;
  var left = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 0;
  var right = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : 0;
  var bottom = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : 0;
  return {
    position: 'absolute',
    top: top,
    right: right,
    bottom: bottom,
    left: left
  };
};

exports.__linariaPreval = [fill1, fill2];"
`;

exports[`keeps only code which is related to \`a\` 1`] = `
"\\"use strict\\";

var _require = require('…'),
    color = _require.whiteColor,
    anotherColor = _require.anotherColor;

var a = color || anotherColor;
color.green = '#0f0';
exports.__linariaPreval = [a];"
`;

exports[`keeps only code which is related to \`anotherColor\` 1`] = `
"\\"use strict\\";

var _require = require('…'),
    anotherColor = _require.anotherColor;

exports.__linariaPreval = [anotherColor];"
`;

exports[`keeps only code which is related to \`color\` 1`] = `
"\\"use strict\\";

var _require = require('…'),
    color = _require.whiteColor;

color.green = '#0f0';
exports.__linariaPreval = [color];"
`;

exports[`keeps reused exports 1`] = `
"\\"use strict\\";

var bar = function bar() {
  return 'hello world';
};

exports.bar = bar;
var foo = exports.bar();
exports.__linariaPreval = [foo];"
`;

exports[`removes all 1`] = `
"\\"use strict\\";

exports.__linariaPreval = [];"
`;

exports[`shakes assignment patterns 1`] = `
"\\"use strict\\";

var _interopRequireDefault = require(\\"@babel/runtime/helpers/interopRequireDefault\\");

var _toArray2 = _interopRequireDefault(require(\\"@babel/runtime/helpers/toArray\\"));

var _extends2 = _interopRequireDefault(require(\\"@babel/runtime/helpers/extends\\"));

var _ = 2,
    identifier = _ === void 0 ? 1 : _;
var _a$b = {
  a: 1,
  b: 2
};
_a$b = _a$b === void 0 ? {} : _a$b;
var object = (_extends2.default)({}, _a$b);
var _ref = [1, 2, 3, 4];
_ref = _ref === void 0 ? [] : _ref;

var _ref2 = (_toArray2.default)(_ref),
    array = _ref2.slice(0);

var obj = {
  member: null
};
var _2 = 1;
obj.member = _2 === void 0 ? 42 : _2;
exports.__linariaPreval = [identifier, object, array, obj];"
`;

exports[`shakes es5 exports 1`] = `
"\\"use strict\\";

Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
exports.redColor = 'red';
exports['yellowColor'] = 'yellow';
Object.defineProperty(exports, \\"greenColor\\", {
  enumerable: true,
  get: function get() {
    return 'green';
  }
});"
`;

exports[`shakes exports 1`] = `
"\\"use strict\\";

var _ = require(\\"\\\\u2026\\");

Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
var a = _.whiteColor;
exports.__linariaPreval = [a];"
`;

exports[`shakes for-in statements 1`] = `
"\\"use strict\\";

var obj1 = {
  a: 1,
  b: 2
};
var obj2 = {};

for (var key in obj1) {
  obj2[key] = obj1[key];
}

exports.__linariaPreval = [obj2];"
`;

exports[`shakes imports 1`] = `
"\\"use strict\\";

var _typeof = require(\\"@babel/runtime/helpers/typeof\\");

Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});

var _ = _interopRequireWildcard(require(\\"\\\\u2026\\"));

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== \\"function\\") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== \\"object\\" && typeof obj !== \\"function\\") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== \\"default\\" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

exports.__linariaPreval = [_.whiteColor, _.default];"
`;

exports[`shakes sequence expression 1`] = `
"\\"use strict\\";

var color1 = (function () {
  return 'blue';
});
var local = '';
var color2 = (local = color1(), function () {
  return local;
});
exports.__linariaPreval = [color2];"
`;

exports[`should keep member expression key 1`] = `
"\\"use strict\\";

var key = 'blue';
var obj = {
  blue: '#00F'
};
var blue = obj[key];
exports.__linariaPreval = [blue];"
`;
